### PR TITLE
extmod/modonewire.c: Fix onewire timing and wait for clear bus.

### PR DESCRIPTION
**Pull request by xorbit**
The timing of the onewire module was way too fast when reading.
The write timing was correct, with a 10us low pulse for 1 and
60us pulse for 0. But read would pulse low for 5us and then sample
the wire after 10us. This is way too fast since a 1 can be
a pulse of up to 15us and would be read as 0. Plus the whole bit
time for read was only 50us, while a client may output a pulse
of up to 60us to indicate a 0. Compounding the problem, the code
would not check if the bus was high before starting a bit pulse,
just blindly plowing through whatever the slave might have been
sending with the next bit before the slave was done.

This code now outputs a 10us pulse for read to be consistent with
the write, then samples the bus after 30us as is the case with
most implementations I found. The total read bit time is 60us.
The code also first checks whether the bus is high before starting
a bit, and waits in 10us increments up to 100us before starting to
output a bit pulse.

### CODE REVIEW

**Comment by xorbit** [comment]
Not sure what the coverage failure is all about, when clicking on details it doesn't show any change in `/extmod` where my change was made.

**Comment by tve** [comment]
What is the reference you use for the protocol timing spec? I'm looking at a DS18B20-PAR datasheet and I see:

"A read time slot is initiated by the master device pulling the 1-Wire bus low for a minimum of 1 μs and then releasing the bus."
"Output data from the DS18B20-PAR is valid for 15 μs after the falling edge that initiated the read time slot. Therefore, the master must release the bus and then sample the bus state within 15 μs from the start of the slot."
"Figure 14 illustrates that the sum of T INIT , T RC , and T SAMPLE must be less than 15 μs for a read time slot."
So IMHO the read timing is fine. Checking that the line is high before initiating a bit transfer makes sense, but I have net seen any driver that does that, and I've seen and coded my fair share of one-wire drivers...

(This was started as PR #7049 and there are comments there too)

**Comment by xorbit** [comment]
You are correct.  I think I misread the Wikipedia article: https://en.wikipedia.org/wiki/1-Wire
It is a bit odd that it says the master read pulse can be up to 15 μs.  Would not be the best idea to also sample at that time if you had only just released the bus.  I think I misinterpreted the slave's 30 μs sample time and assumed the master would do the same to position the sample somewhat central in the 15 μs to 60 μs window.  I was working with a software implemented slave, not a hardware chip.

I also had referenced an Arduino library on Github but for the life of me I can't find it anymore now.  Maybe it was removed because it was wrong. :)  It was not the one by Paul Stoffregen (https://github.com/PaulStoffregen/OneWire/blob/master/OneWire.cpp) that I now prominently find, which pulses 3 μs low, then samples after 10 μs (so sample time is 13 μs from falling edge) and then waits 53 μs for a total bit time of 66 μs.

In the datasheet you reference: https://datasheets.maximintegrated.com/en/ds/DS18B20-PAR.pdf
Indeed they say starting on the bottom of page 14 that the output is valid for 15 μs after the falling edge.  Then they recommend doing it toward the end of the 15 μs because of rise time.  However, they also say: "All  read  time  slots  must  be  a  minimum  of  60  μs  in  duration  with  a  minimum  of  a  1  μs  recovery  time  between  slots."  The existing code only provides a 50 μs read slot so in that respect the read time is not fine.

Then there's this document from Maxim: https://www.maximintegrated.com/en/design/technical-documents/app-notes/1/126.html
It recommends a 6 μs read pulse from the master, then sampling 9 μs later.  So the sample time is 15 μs from the falling edge.

The DS18B20-PAR datasheet on page 19 shows a timing diagram that illustrates the slop on the inaccurate RC timing of the 0 pulse which can range from 15 μs to Tslot, which is specified as ranging from 60 μs to 120 μs.  Since it can be that long, I think keeping the wait for the bus to go high before a bit transfer would be a good idea.

So I'm thinking of making these changes:

``` python
#define TIMING_READ1  (6)
#define TIMING_READ2  (9)
#define TIMING_READ3  (55)
```

This would match the timings specified in this Maxim document: https://www.maximintegrated.com/en/design/technical-documents/app-notes/1/126.html
What do you think?

**Comment by tve** [comment]
I guess the question that comes to mind is: what is the problem you're trying to solve?

**Comment by xorbit** [comment]
The problem I'm trying to solve is that I'm trying to make it work.  With more devices.

The current code doesn't work with slave devices that happen to generate a low longer than 50 μs for 0, something they are very much allowed to do.  The reason is that the current master code doesn't respect the bit slot time of 60 μs. It also just starts the next bit regardless of whether the slave is done sending the previous bit or not. And as shown in the datasheet you referred to, the slot time can actually range between 60 and 120 μs.

Sampling at 10 μs is also quite early, Maxim's recommendation is 15 μs according to the app note I linked (https://www.maximintegrated.com/en/design/technical-documents/app-notes/1/126.html). Because the rising edge is generated by a pull-up, if you have a long bus with many devices, the bus voltage may not rise quickly enough to be read correctly at 10 μs.  This is a slow protocol, it requires you to wait the while bit slot time of 60 μs anyway, so there really isn't a good reason to sample the bus earlier than necessary.  Giving it that extra time can make it work in situations it can't right now.

So you could say "the current code works for me", and that may be true.  But it doesn't follow the spec for bit slot time, or the recommended timing from Maxim.  So it may work for your device, sitting on your desk, but by improving it, it may work in the field with longer buses, more devices, and other devices.  That would be a good problem to solve in my book.

And more to the point: I am currently working with a software slave that does everything within spec, but may be slower than a hardware device.  It doesn't work with the existing code, it does work with the latest code in this pull request.

**Comment by robert-hh** [comment]
Do you know this analysis: https://github.com/cpetrich/counterfeit_DS18B20

**Comment by xorbit** [comment]
@robert-hh No, I didn't.  It is interesting but I don't really see the relevance to this PR.  I'm not using a DS18B20, and the issue I want to fix is a violation of the Maxim spec.

**Comment by tve** [comment]
In my experience, if you're having difficulties with 1-wire buses that have lots of devices or long wires you're better off looking at the signal integrity than tweaking timing. For example, I generally put a 180 ohm serial resistor to dampen reflections and use a rather low value pull-up resistor. Also, using a slow slew rate (low power drive) on the uC pin can help a lot. For tough situations, using a ds2483 can save a ton of headaches...

Which port are you using and have you looked at the timing of signals with a scope? A delay of 5us plus a bunch of function calls can easy account for 6us total, for example.

**Comment by xorbit** [comment]
You seem to be under the mistaken impression that I'm trying to fix "my problem" here.  I don't have a problem, I was just using MicroPython as a test vehicle, found the onewire implementation lacking, fixed it, did what I needed to do with it and decided to contribute the fix back because the implementation as it is does not comply with the spec.

I'm using the Raspberry Pi Pico with the MicroPython RP2 port.  Below is a scope trace showing a read with the 1.14 release code:
Newfile1
As you can see, it works with this particular slave because it has a 0 pulse of about 40 μs.  But you can imagine what would happen if this happened to be a slave that output a 60 μs pulse, which is perfectly in spec: the next bit read pulse would arrive _while_ the previous slave output was still active and everything would go to hell.

Here it is with my fixed code:
Newfile2
It respects the bit slot time, following the timing recommended by Maxim.  If this happened to be a slave with 60 μs 0 pulse, it would work.  If this happened to be a slave with even longer bit slot time (up to 120 μs according to the DS18B20-PAR spec), the code would wait up to 100 μs longer before outputting the next bit pulse, and it would still work.

Here is the read pulse zoomed in:
Newfile3
Looks like your worry about function calls making it longer is unfounded, it's actually a little less than 6 μs.

There is actually another aspect of this you have to keep in mind: if you are using slaves that run from parasitic power, the longer bit slot time is also beneficial because it gives you more "high time" to keep the slave power up.

**Comment by tve** [comment]
I took a look at the DS2483 I2C 1-wire bus master datasheet. The std timing (table 17) is a 64us slot pus a 5.25us recovery time, i.e. basically 70us iteration time. I could not find anything in the description of the chip that would cause it to wait longer than those 70us.

Looking at the ds18b20 datasheet, the max 120us for Tslot does not mean that the chip may hold the line low for that long. It means that the master needs to keep things moving and not keep the line low for more than 120us or else the ds18B20 may not play ball (it may run out of juice, for example, or go to sleep).

So I can see that it makes sense to adjust the timing values 'cause fast processors (like the rp2) go a bit too fast. It would be interesting to know why the current values were chosen. But I don't see why the loop waiting for the line to go high would be necessary.

**Comment by xorbit** [comment]
I guess we read the datasheet differently.  Here's the definition of tSLOT:
Screenshot from 2021-03-20 12-49-19

And according to this diagram, the rising edge can occur anywhere between tRDV and tSLOT:
Screenshot from 2021-03-20 12-49-09

Having the wait loop doesn't hurt anything if the slave is well behaved and releases the bus within 60 μs.  I assume this will be most cases.  But I don't really see the point in preventing it from working in cases where it may be longer than that.

**Comment by xorbit** [comment]
If there are no more objections, can this be merged please?

**Comment by dpgeorge** [comment]
Thanks for the contribution.

I'm not sure I agree with the new timing values for the read... from the DS18B20-PAR datasheet (linked above), it says that "the master must release the bus and then sample the bus state within 15us from the start of the slot" (p15).  This means that tRDV=15us is the maximum time that the data is valid, not the minimum time to wait for the data to be valid.  So the new timing values here will put the read right at the end of this window (6us+9us) and will probably lead to issues because the actual read may happen just after 15us due to the time taken to do the read.

If anything the existing values could be changed so that `TIMING_READ1=1` and `TIMING_READ2=11` or so (the datasheet says to minimise tINIT to around 1us).

As for the maximum length of the bit slot: I think what is needed is at the end of the readbit function, if the read value was 0 then it should wait (at most 120us) for the line to go high before returning.  No change is needed for writebit.

**Comment by smurfix** [comment]
> it says that "the master must release the bus and then sample the bus state within 15us from the start of the slot" (p15). This means that tRDV=15us is the maximum time that the data is valid, not the minimum time to wait for the data to be valid.

That's basically what the data sheet of the DS2770 says, too.

@xorbit Don't forget that the 1wire bus timing is asymmetric. The master can sample the bus a few µs after it releases it because it bloody well knows that the slave either holds the wire down or not. Thus if you send a 1 µs pulse and then wait another 3 µs before sampling you're perfectly within spec (though if the bus wire is longer and/or unterminated this is not very reliable). As the slave is within its rights to release the wire at any time after those 15µs are up it's dangerous to sample the bus later than that, even though many slaves will hold the wire down for longer (at most 60 µs or so).

The slave, on the other hand, must rely entirely on timing. It samples the bus some unspecified time from 15 µs to 60 µs after the start of the master's pulse; the DS2770's typical value is 30 µs.

**Comment by xorbit** [comment]
This PR has two sides to it:

1. Fix the bit slot time for read which is currently in violation of the spec.
2. Try to make the driver as robust as possible to work with as wide a range of devices as possible (devices implemented in hardware as well as software), and for as wide a range of bus lengths and implementations as possible.

For the first part, that the bit slot time for read is out of spec is without question: it's currently 50 µs while it should be _at least_ 60 µs.  You can say: "but it works for my chip and on my desk" but that's irrelevant, the next chip may not work because you violated the slave spec.

The second part is more open to interpretation.  You can go: "look, this particular slave device says it only needs a 1 µs pulse and you can read the bit right after", and implement the master that way.  But then please, rename the driver to "ds18b20" instead of "onewire" because that's what you've made.  [As a side point, at system level, a 1 µs read pulse would likely cause problems even for DS18B20s if you have a very long bus or some form of slope control.  Don't do that even for a dedicated DS18B20 driver.]

Instead of trying to find the shortest read pulse you could possibly get away with (who cares anyway, the total bit slot should be 60 µs no matter what), my goal instead was to make this robust to work across various devices, some of which were implemented in software.  If the master sends a read bit of 1 µs, then a microcontroller slave likely hasn't even woken up by the time the bit is over, let alone had time to pull the bus low.

Now you may say: "I don't care about supporting slow software devices".  But why wouldn't you?  It doesn't cost you anything if you use master timing that works for both.  Or are you trying to make a "onewire-hardware-device-only" driver?  Shouldn't the goal be to be as compatible and robust as possible?

As for when to sample the bus: even at 15 µs, remember that the high transition is through a pull-up, so even if the sample was taken at the exact same time the bus was released, it would still be read as a 0.

The bottom line is: the read timing in this PR exactly matches what Maxim recommends in this app note entitled [1-Wire Communication Through Software](https://www.maximintegrated.com/en/design/technical-documents/app-notes/1/126.html).  I think that's relevant to what you're doing here?  Even if you don't agree with my, don't you agree they are the authority on how to implement this?

**Comment by villeneuve** [comment]
Hello,
I wanted to open an issue but found this relevant PR still open, so I post my findings here.
I did onewire (ds18b20) tests with a logic analyser. 
When testing devices running MicroPython, [PulseView](https://sigrok.org/wiki/PulseView) reports the read time slots as  "Time slot not long enough" As per this screenshot:
TimeSlotTooShortMicroPythonRaspbPico

This happens with Raspberry Pi pico running latest stable MicroPython v1.17 on 2021-09-02
Same with ESP8266 running latest stable MicroPython v1.17 on 2021-09-02
This doesn't happen with the same board ESP8266 flashed with Arduino software (onewire lib)
Doesn't happen as well with an Arduino Uno.
In these two laters the read time slots are just above 60us when it's around 50us with MicroPython.
PulseView 1wire-link decoder can decode the signal of the Arduino programmed devices but not with MicroPython.

I dig in the Maxim 1wire data sheets and found that this sentence:
1 - All read time slots must be a minimum of 60µs in duration with a minimum of a 1µs recovery timebetween slots
2 - And this table "AC electrical characteristics"  showing Tslot Min and Max (already mentioned by xorbit) 
Screenshot 2021-09-29 17 46 34

Are common to many data sheet:
[DS18B20-PAR](https://datasheets.maximintegrated.com/en/ds/DS18B20-PAR.pdf) : first point page 14 second point page 18
[DS18B20](https://datasheets.maximintegrated.com/en/ds/DS18B20.pdf) : first point page 16 second point page 3
[DS18B22](https://datasheets.maximintegrated.com/en/ds/DS1822.pdf): first point page 16 second point page 20
[DS18B25](https://datasheets.maximintegrated.com/en/ds/DS1825.pdf): first point page 17 second point page 3
[DS18S20](https://datasheets.maximintegrated.com/en/ds/DS18S20.pdf): first point page 16 second point page 3
And also the [Book of iButton standard ](https://pdfserv.maximintegrated.com/en/an/AN937.pdf)p40 mentions Tslot min 60 max 120us

So it seems clear that the manufacturer specification of the read time slot is: **between 60 and 120µs.**

**Comment by xorbit** [comment]
Sorry, by sitting idle, this PR seems to have collected a couple of unrelated files.  Only `modonewire.c` should be in the PR, but I don't have write permission to remove the other ones.
If this had been promptly merged that wouldn't have happened of course. ;)  Not sure why it's so hard to get a fix for an obvious bug merged.

**Comment by dpgeorge** [comment]
After reading through https://www.analog.com/en/technical-articles/1wire-communication-through-software.html, I agree that the current timings for read-bit should be changed to match the reference values in that linked document.  For sure the current read-bit is way too short (50us per bit when it should be 70us per that document).

The write-bit timings can also be tuned, see #13171.

**Comment by dpgeorge** [comment]
In 06df3b29256e4e90abc61505a28852268cf64af7 I merged the commit to adjust the read timings.  With this, I don't think it's necessary to wait for the bus to be high, so I did not merge that bit.
